### DISCUSSION TITLE
upgrade pulsar 2.7.x to 2.10.x: curator and zookeeper compatibility problem

### DISCUSSION THREAD

**Discussion by Maslino** [apache/pulsar#19709] [Q&A]
Pulsar 2.7.x uses curator 4 and zookeeper 3.5.x
From Pulsar 2.8.x onwards, pulsar uses curator 5 and zookeeper 3.6.x

I want to upgrade pulsar 2.7.x to 2.10.x, but face the problem of compatibility between curator library and zookeeper server.

Following the upgrade guidelines from pulsar docs, the upgrade sequence is firstly zookeeper then bookkeeper and then broker. And zookeeper upgrade is optional.

If I choose to not to upgrade zookeeper: are bookkeeper and broker of pulsar 2.10.x compatible with zookeeper 3.5.x of pulsar 2.7.x?
If I choose to upgrade zookeeper: is bookkeeper of pulsar 2.7.x compatible with zookeeper 3.6.x of pulsar 3.10.x?

As far as I know, pulsar 2.10.x is not compatible with zookeeper 3.5.x

Is pulsar 2.7.x compatible with zookeeper 3.6.x ?

**Comment by hezhangjian** [comment]
I think it will works

**Comment by tisonkun** [comment]
I don't remember that we use any new features of ZK/Curator on the Pulsar side, so such ZK/Curator version shift should all work.

cc @eolivelli @merlimat

**Reply by tisonkun** [reply]
BTW, perhaps a compatible matrix helps?

**Comment by Maslino** [comment]
[issue #14675](https://github.com/apache/pulsar/issues/14675) shows that pulsar 2.10.x is not compatible with ZK 3.5.x

When I upgrade pulsar 2.7.4 to 2.10.3 without ZK upgrade, bookie seems OK but broker throws the same exception (**UnimplementedException: KeeperErrorCode = Unimplemented for /**) as described in the above issue failing to start.

**Reply by tisonkun** [reply]
@Maslino I think it's because the zk client in broker classpath is >= 3.6.3. You may try to replace the zookeeper JARs under `$PULSAR_HOME/lib` with 3.5.x version and see what happens.

Curator conditionally uses new features based on what version of ZK in the classpath.

**Reply by Maslino** [reply]
I am using the official pulsar-all docker image.

**Reply by tisonkun** [reply]
Well. Then it can be a problem 🤣

**Reply by hezhangjian** [reply]
@Maslino would you mind rebuild a image to replace zookeeper library inside lib directory?
